braindecode/models/conformer.py

```diff
@@ -0,0 +1,279 @@
+                kernel_avg_pool,
+                stride_avg_pool,
+                conv_drop,
+            ),
+            TransformerEncoder(att_depth, kernel, att_heads, att_drop),
+            ClassificationHead(kernel, fc_dim, n_classes),
+        )
+
+
+class PatchEmbedding(nn.Module):
+    """Patch Embedding.
+
+    The authors used a convolution moduleto capture local features,
+    instead of postion embedding.
+
+    """
+
+    def __init__(
+            self,
+            kernel,
+            kernel_temp_conv,
+            eeg_channel,
+            kernel_avg_pool,
+            stride_avg_pool,
+            conv_drop,
+    ):
+        super().__init__()
+
+        self.shallownet = nn.Sequential(
+            nn.Conv2d(1, kernel, (1, kernel_temp_conv), (1, 1)),
+            nn.Conv2d(kernel, kernel, (eeg_channel, 1), (1, 1)),
+            nn.BatchNorm2d(kernel),
+            nn.ELU(),
+            nn.AvgPool2d(
+                (1, kernel_avg_pool), (1, stride_avg_pool)
+            ),
+            # pooling acts as slicing to obtain 'patch' along the
+            # time dimension as in ViT
+            nn.Dropout(conv_drop),
+        )
+
+        self.projection = nn.Sequential(
+            nn.Conv2d(
+                kernel, kernel, (1, 1), stride=(1, 1)
+            ),  # transpose, conv could enhance fiting ability slightly
+            Rearrange("b e (h) (w) -> b (h w) e"),
+        )
+
+    def forward(self, x: Tensor) -> Tensor:
+        x = x.unsqueeze(dim=1)  # add one extra dimension
+        x = self.shallownet(x)
+        x = self.projection(x)
+        return x
+
+
+class MultiHeadAttention(nn.Module):
+    def __init__(self, emb_size, num_heads, dropout):
+        super().__init__()
+        self.emb_size = emb_size
+        self.num_heads = num_heads
+        self.keys = nn.Linear(emb_size, emb_size)
+        self.queries = nn.Linear(emb_size, emb_size)
+        self.values = nn.Linear(emb_size, emb_size)
+        self.att_drop = nn.Dropout(dropout)
+        self.projection = nn.Linear(emb_size, emb_size)
+
+    def forward(self, x: Tensor, mask: Tensor = None) -> Tensor:
+        queries = rearrange(
+            self.queries(x), "b n (h d) -> b h n d", h=self.num_heads
+        )
+        keys = rearrange(
+            self.keys(x), "b n (h d) -> b h n d", h=self.num_heads
+        )
+        values = rearrange(
+            self.values(x), "b n (h d) -> b h n d", h=self.num_heads
+        )
+        energy = torch.einsum("bhqd, bhkd -> bhqk", queries, keys)
+        if mask is not None:
+            fill_value = torch.finfo(torch.float32).min
+            energy.mask_fill(~mask, fill_value)
+
+        scaling = self.emb_size ** (1 / 2)
+        att = F.softmax(energy / scaling, dim=-1)
+        att = self.att_drop(att)
+        out = torch.einsum("bhal, bhlv -> bhav ", att, values)
+        out = rearrange(out, "b h n d -> b n (h d)")
+        out = self.projection(out)
+        return out
```

Comment:
Why do you not use [nn.MultiheadAttention](https://pytorch.org/docs/stable/generated/torch.nn.MultiheadAttention.html)?

Reply:
Hello， @martinwimpff and @bruAristimunha
Thanks for your help!
I've compared torch implementation and manual implementation (running two programs on two same GPUs). The memory (~7G->9G) and time consumption (12min -> 13 min per subject) were raised with the torch version. The accuracy decreased by 4.xx% on the BCI competition IV 2a dataset.
So I suggest using the original implementation, which could also help users to figure out the network design. 

If you still prefer the torch version, I may need more time to compare the difference in detail.

Reply:
Can you please check the number of parameters? Just to see if both methods yield the same architecture. 

One difference that I know of is the weight initialization of `nn.MultiheadAttention` vs. the weight initialization of your `nn.Linear` layers (xavier_uniform vs. kaiming_uniform). So what you could also check is whether the acc decrease is just coming from this weight init.

Strange to see that the torch version needs more memory as it usually should be better optimized and therefore more efficient.

Reply:
Hey @martinwimpff and @eeyhsong,

To speed it up a bit, can we address this issue in another issue/pull request? I think this question can be extended to other models in the library.

Reply:
It will handle in #463.